[Docs] Resilience pipeline registry (#1575)
App-vNext/Polly: docs/resilience-pipeline-registry.md

{
    // Define your pipeline
    builder.AddRetry(new RetryStrategyOptions<HttpResponseMessage>());
});
```
<!-- endSnippet -->

## Registry options

The constructor for `ResiliencePipelineRegistry<TKey>` accepts a parameter of type `ResiliencePipelineRegistryOptions<TKey>`. This parameter lets you configure the behavior of the registry. Here's a breakdown of the available properties:

| Property                | Default Value                                                   | Description                                                       |
| ----------------------- | --------------------------------------------------------------- | ----------------------------------------------------------------- |
| `BuilderFactory`        | Function returning a new `ResiliencePipelineBuilder` each time. | Allows consumers to customize builder creation.                   |
| `PipelineComparer`      | `EqualityComparer<TKey>.Default`                                | Comparer the registry uses to fetch resilience pipelines.         |
| `BuilderComparer`       | `EqualityComparer<TKey>.Default`                                | Comparer the registry uses to fetch registered pipeline builders. |

> | `InstanceNameFormatter` | `null`                                                          | Function formatting `TKey` to instance name.                      |
> | `BuilderNameFormatter`  | Function returning the `key.ToString()` value.                  | Function formatting `TKey` to builder name.                       |

Reviewer: ```suggestion
| `InstanceNameFormatter` | `null`                                                          | Delegate for formatting `TKey` to instance name.                      |
| `BuilderNameFormatter`  | Delegate returning the `key.ToString()` value.                  | Delegate for formatting `TKey` to builder name.                       |
```

// Dispose the registry
registry.Dispose();

try
{
    pipeline.Execute(() => { });
}
catch (ObjectDisposedException)
{
    // Using a pipeline that was disposed by the registry
}
```
<!-- endSnippet -->

The registry also allows for the registration of dispose callbacks. These are called when a pipeline is discarded, either because of the registry's disposal or after the pipeline is reloaded. The example below works well with dynamic reloads, letting you dispose of the `CancellationTokenSource` when it's not needed anymore.

<!-- snippet: registry-reloads-and-dispose -->
```cs
var registry = new ResiliencePipelineRegistry<string>();

registry.TryAddBuilder("A", (builder, context) =>
{
    var cancellation = new CancellationTokenSource();

    // Register the source for potential external triggering
    RegisterCancellationSource(cancellation);

    // Add the reload token; note that an already cancelled token is disregarded
    context.AddReloadToken(cancellation.Token);

    // Configure your pipeline
    builder.AddRetry(new RetryStrategyOptions());

    context.OnPipelineDisposed(() => cancellation.Dispose());
});
```
<!-- endSnippet -->

> Both `AddReloadToken(...)` and `OnPipelineDisposed(...)` are used to implement the `EnableReloads<TOptions>(...)` extension method that is used by the [DI layer](dependency-injection.md#dynamic-reloads).

Reviewer: ```suggestion
Both `AddReloadToken(...)` and `OnPipelineDisposed(...)` are used to implement the `EnableReloads<TOptions>(...)` extension method that is used by the [Dependency Injection layer](dependency-injection.md#dynamic-reloads).
```

| `BuilderFactory`        | Function returning a new `ResiliencePipelineBuilder` each time. | Allows consumers to customize builder creation.                   |
| `PipelineComparer`      | `EqualityComparer<TKey>.Default`                                | Comparer the registry uses to fetch resilience pipelines.         |
| `BuilderComparer`       | `EqualityComparer<TKey>.Default`                                | Comparer the registry uses to fetch registered pipeline builders. |
| `InstanceNameFormatter` | `null`                                                          | Delegate formatting `TKey` to instance name.                      |

> | `BuilderNameFormatter`  | Function returning the `key.ToString()` value.                  | Delegate formatting `TKey` to builder name.                       |

Reviewer: ```suggestion
| `BuilderNameFormatter`  | A delegate returning the `key.ToString()` value.                  | Delegate formatting `TKey` to builder name.                       |
```
